x/interchainqueries/keeper/keeper.go

```diff
@@ -132,9 +133,59 @@ func (k Keeper) GetAllRegisteredQueries(ctx sdk.Context) []*types.RegisteredQuer
 	return queries
 }
 
-func (k Keeper) RemoveQueryByID(ctx sdk.Context, id uint64) {
+// RemoveQuery removes the given query and relative result data from the store. For a KV query it
+// deletes the *types.QueryResult stored by the query ID, for a TX query it stores the query ID to
+// the list of queries to be removed so the ICQ module can remove the query hashes later.
+func (k Keeper) RemoveQuery(ctx sdk.Context, query *types.RegisteredQuery) {
 	store := ctx.KVStore(k.storeKey)
-	store.Delete(types.GetRegisteredQueryByIDKey(id))
+	store.Delete(types.GetRegisteredQueryByIDKey(query.Id))
+	queryType := types.InterchainQueryType(query.GetQueryType())
+	switch {
+	case queryType.IsKV():
+		store.Delete(types.GetRegisteredQueryResultByIDKey(query.Id))
+	case queryType.IsTX():
+		store.Set(types.GetTxQueryToRemoveByIDKey(query.Id), []byte{})
+	}
+}
+
+// TxQueriesCleanup cleans the module store from obsolete registered TX queries and relative
+// stored transaction hashes. Cleans up to params.TxQueryRemovalLimit hashes at a time or all
+// the hashes if params.TxQueryRemovalLimit is 0.
+func (k Keeper) TxQueriesCleanup(ctx sdk.Context) {
+	st := time.Now()
+	rmLimit := k.GetParams(ctx).TxQueryRemovalLimit
+	limited := rmLimit != 0
+
+	queriesToRm := make([]*TxQueryToRemove, 0, rmLimit/10)
+	for _, queryID := range k.GetTxQueriesToRemove(ctx, rmLimit) {
+		queryToRm := k.calculateTxQueryRemoval(ctx, queryID, rmLimit)
+		queriesToRm = append(queriesToRm, queryToRm)
+
+		if limited {
+			rmLimit -= uint64(len(queryToRm.Hashes))
+			if rmLimit <= 0 {
+				break
+			}
+		}
+	}
```

Comment:
Why don't you want to use nested loop here? It will reduce memory/resource consumption because of lesser amount of iterations.
Something like
for TxQueriesToRemove.Iterator {
   queryToRm := k.calculateTxQueryRemoval(ctx, queryID, rmLimit)
   queriesToRm = append(queriesToRm, queryToRm)
}

Reply:
I tried it but it didn't work: there are iterators over the same resource inside both `calculateTxQueryRemoval` and `GetTxQueriesToRemove`, and the iterator is tricky designed so one of them got blocked even without write operations. i.e. when I iterated inside another iterator loop I got deadlocked. idk why but this is why I've chosen such an approach